options/posix/generic/pthread-stubs.cpp

```diff
@@ -790,7 +790,11 @@ int pthread_mutex_lock(pthread_mutex_t *mutex) {
 
 			// Wait on the futex if the waiters flag is set.
 			if(expected & mutex_waiters_bit) {
-				if(int e = mlibc::sys_futex_wait((int *)&mutex->__mlibc_state, expected, nullptr); e)
+				int e = mlibc::sys_futex_wait((int *)&mutex->__mlibc_state, expected, nullptr);
+
+				// If the wait returns EAGAIN, that means that the mutex_waiters_bit was just unset by
+				// some other thread. In this case, we should loop back around.
+				if (e && e != EAGAIN)
```

Comment:
Why the refactor to separate out this variable? (also is there no real mechanism for reporting this error out of libc?)

Reply:
how else would it be written

Reply:
I think this is the right fix. The code is called in a loop and EAGAIN simply means that we need to loop once more because there was a race (the futex state did not match our expectation).

Reply:
just as before: `if (int e = mlibc::sys_futex_wait((int *)&mutex->__mlibc_state, expected, nullptr); e && e != EAGAIN)`

> I think this is the right fix. The code is called in a loop and EAGAIN simply means that we need to loop once more because there was a race (the futex state did not match our expectation).

I mean non-EAGAIN errors, my bad for the confusion.

Reply:
True, just the line is really long that way. I don't think there is anything more we can do with the error here; at least the man page has a list of only a few errors lock can return.